tensorflow/lite/micro/kernels/exp.cc

```diff
@@ -12,78 +12,69 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 ==============================================================================*/
+
+#include "tensorflow/lite/kernels/internal/reference/exp.h"
+
 #include "tensorflow/lite/c/common.h"
-#include "tensorflow/lite/kernels/internal/reference/reference_ops.h"
-#include "tensorflow/lite/kernels/internal/tensor.h"
 #include "tensorflow/lite/kernels/internal/tensor_ctypes.h"
 #include "tensorflow/lite/kernels/kernel_util.h"
+#include "tensorflow/lite/micro/kernels/kernel_util.h"
 
 namespace tflite {
 namespace ops {
-namespace builtin {
+namespace micro {
 namespace exp {
 
-// This file has reference implementation of Exp.
-enum KernelType {
-  kReference,
-};
-
-struct ExpContext {
-  ExpContext(TfLiteContext* context, TfLiteNode* node) {
-    input = GetInput(context, node, 0);
-    output = GetOutput(context, node, 0);
-  }
-  const TfLiteTensor* input;
-  TfLiteTensor* output;
-};
-
 TfLiteStatus Prepare(TfLiteContext* context, TfLiteNode* node) {
   TF_LITE_ENSURE_EQ(context, NumInputs(node), 1);
   TF_LITE_ENSURE_EQ(context, NumOutputs(node), 1);
-
-  ExpContext op_context(context, node);
-  TfLiteIntArray* output_dims = TfLiteIntArrayCopy(op_context.input->dims);
-  op_context.output->type = op_context.input->type;
-  return context->ResizeTensor(context, op_context.output, output_dims);
+  const TfLiteTensor* input = GetInput(context, node, 0);
+  TF_LITE_ENSURE(context, input != nullptr);
+  TfLiteTensor* output = GetOutput(context, node, 0);
+  TF_LITE_ENSURE(context, output != nullptr);
+  TF_LITE_ENSURE_TYPES_EQ(context, input->type, kTfLiteFloat32);
+  TF_LITE_ENSURE_TYPES_EQ(context, output->type, input->type);
+  TF_LITE_ENSURE_EQ(context, output->bytes, input->bytes);
+  TF_LITE_ENSURE_EQ(context, output->dims->size, input->dims->size);
+  for (int i = 0; i < output->dims->size; ++i) {
+    TF_LITE_ENSURE_EQ(context, output->dims->data[i], input->dims->data[i]);
+  }
+  return kTfLiteOk;
 }
 
-template <KernelType kernel_type>
 TfLiteStatus Eval(TfLiteContext* context, TfLiteNode* node) {
-  ExpContext op_context(context, node);
-
-#define TF_LITE_EXP(kernel_type, data_type)                               \
-  kernel_type::Exp<data_type>(GetTensorData<data_type>(op_context.input), \
-                              NumElements(op_context.input),              \
-                              GetTensorData<data_type>(op_context.output))
+  const TfLiteEvalTensor* input =
+      tflite::micro::GetEvalInput(context, node, 0);
+  TfLiteEvalTensor* output =
+      tflite::micro::GetEvalOutput(context, node, 0);
```

Comment:
pull the magic constants out to a constexpr.

https://github.com/tensorflow/tensorflow/blob/b0fcce8080f94b44195c429929361a420e9255cc/tensorflow/lite/micro/kernels/conv.cc#L31-L34

Reply:
fixed.